### PR TITLE
Upgrade bitcoin dependency to v0.29.0 

### PR DESCRIPTION
**Pull request by tcharding**
This is going to need a pretty careful review by @sanket1729 and/or @apoelstra to double check all the sequence/lock_time stuff. Even with the new APIs this stuff is non-trivial. This patch changes the current behaviour around logic for these two types. I am unable to work out if current behaviour has bugs in it or if I just don't fully understand it. 

### Note

Please ignore the branch name, it stale now since this PR upgrades to the _newly released_ bitcoin v0.29.1

### CODE REVIEW

File: src/miniscript/types/mod.rs
Comment (Kixunil) [inline src/miniscript/types/mod.rs:824]: Did the code originally attempt to check if lock times are actually enabled or something?

**Review by Kixunil — CHANGES_REQUESTED** [review]
I found a few mistakes. Disclaimer: didn't do super-deep review.

File: src/interpreter/mod.rs
Comment (tcharding) [reply src/interpreter/mod.rs:499]: Its a block height or a block time. `n` is from `n OP_CLTV` and `nLockTime` - that's why I used `n`.

Comment (Kixunil) [reply src/interpreter/mod.rs:499]: `n` in `nLockTime` literally means number. It's meaningless. I do see that `time` is confusing, so `lock_time` would be better because most people already know it can be height.

File: src/miniscript/hash256.rs
Comment (tcharding) [reply src/miniscript/hash256.rs:46]: I added this because of the trait bounds on `Hash` trait (from `bitcoin_consensus`). `Deref` doesn't help with that, right? Or am I missing something?

File: src/miniscript/types/mod.rs
Comment (tcharding) [reply src/miniscript/types/mod.rs:824]: Originally both sequnece and locktime (`After` and `Older`) where handled together. When I split them apart I kept the disable check even though it seemed wrong, finally I got the confidence to remove it :)

File: src/miniscript/hash256.rs
Comment (Kixunil) [reply src/miniscript/hash256.rs:46]: I don't see any `Hash` trait in `bitcoin_consensus`, I guess you mean `bitcoin_hashes`. You're right that `Deref` doesn't help but it's a question if that bound should be there since it's quite burdensome. I'd have to look into it deeper. The first check is will anything break if we remove it?

File: src/descriptor/tr.rs
Comment (tcharding) [reply src/descriptor/tr.rs:260]: This whole error match is strange, every error variant has an `unreachable!`. To be honest I was mainly just interested in checking that `rust-bitcoin` worked so I just hacked this up to make it build.

Comment (Kixunil) [reply src/descriptor/tr.rs:260]: They have explanation why it should be OK, so I think it's fine. But it makes me question the whole API. Maybe we should've used typestate instead?

File: src/interpreter/error.rs
Comment (tcharding) [reply src/interpreter/error.rs:192]: face palm :) Cheers

File: src/interpreter/error.rs
Comment (tcharding) [reply src/interpreter/error.rs:240]: Fixed.

**Comment by tcharding** [comment]
Thanks for the review @Kixunil!

File: src/descriptor/tr.rs
Comment (sanket1729) [reply src/descriptor/tr.rs:260]: I will suggest some changes for the upstream API to make this clean

Comment (sanket1729) [reply src/descriptor/tr.rs:260]: Opened https://github.com/rust-bitcoin/rust-bitcoin/pull/1151

File: src/miniscript/types/mod.rs
Comment (apoelstra) [reply src/miniscript/types/mod.rs:824]: It's not just after vs older -- the original code would look at a `u32` which might contain a value that didn't encode any locktime at all.

It's not clear to me that this check can be removed.

Comment (tcharding) [reply src/miniscript/types/mod.rs:824]: No worries, since this PR is not a merge candidate this bridge doesn't need crossing now. I imagine between @apoelstra and @sanket1729 this will have to be worked out when miniscript upgrades to use rust-bitcoin 0.29.0.

File: src/descriptor/tr.rs
Comment (tcharding) [reply src/descriptor/tr.rs]: This is fixed now, please check my use of a single `unreachable` statement @sanket1729

File: src/interpreter/mod.rs
Comment (tcharding) [reply src/interpreter/mod.rs:499]: Used `lock_time` for this field and similarly `sequence` for `SatisfiedConstraint::RelativeTimelock` field.

File: src/miniscript/hash256.rs
Comment (tcharding) [reply src/miniscript/hash256.rs:46]: I'll investigate this, cheers.

File: src/miniscript/types/mod.rs
Comment (tcharding) [reply src/miniscript/types/mod.rs:824]: Flagging this thread, this will need resolving before we can consider upgrading to the newly release version of `rust-bitcoin`. Sorry to be a drag but I think this is kind of subtle and will need input by @sanket1729 and/or @apoelstra to work it out. I can code up the solution.

File: src/miniscript/hash256.rs
Comment (tcharding) [reply src/miniscript/hash256.rs:46]: Please see: https://github.com/rust-bitcoin/bitcoin_hashes/pull/168

File: src/policy/semantic.rs
Comment (sanket1729) [inline src/policy/semantic.rs:599]: Indeed, sorry I was absent during the review of the initial code. I get it that there was some hesitance in implementing `PartialOrd`, but why not have a method that returns Option<cmp::Ordering>? 

We should really not need to check that they are same units here.

File: src/psbt/mod.rs
Comment (sanket1729) [inline src/psbt/mod.rs:354]: I like the other mthods. They make the code here a lot cleaner

**Review by sanket1729 — COMMENTED** [review]
Mostly ACK. did a round of careful review. Just waiting on reply for comparison FIXME comment

File: src/policy/semantic.rs
Comment (apoelstra) [reply src/policy/semantic.rs:599]: `PartialOrd` gives us a method that returns `Option<cmp::Ordering>`. It's `Ord` that we couldn't implement, because there is no natural total ordering on timelocks, but which we wanted to implement so that you could use `Vec::sort` and derive `Ord` on structures that contain timelocks.

Comment (Kixunil) [reply src/policy/semantic.rs:599]: I think @sanket1729 means that `PartialOrd` still allows `<` and `>` operators which are foot guns. Maybe this should be a `clippy` lint.

**Comment by tcharding** [comment]
Thanks for the review @sanket1729, I'll get back to this soonish, no rush because AFAIU there are a few other crates to upgrade first. Feel free to take this work and use it if you don't want to wait for me.

File: src/miniscript/satisfy.rs
Comment (LLFourn) [inline src/miniscript/satisfy.rs:131]: The previous implementation masked the sequence before doing the comparison. How can we justify not doing this here? What if one has the RBF disabled  and the other doesn't for example (disabling RBF will always be the highest value unless you mask out that bit).

PS methods on `Sequence` like `is_satisfied_by_height` and `is_satisfied_by_time` would be super useful here.

Comment (tcharding) [reply src/miniscript/satisfy.rs:131]: > How can we justify not doing this here?

Thanks man, you are 100% correct, this code is not the same as the code it replaces - that's my bug.

> PS methods on Sequence like is_satisfied_by_height and is_satisfied_by_time would be super useful here.

A relative lock time type is meant to be in the works (not sure if anyone is actively working on it yet) and that type would have these methods you suggest. When `Sequence` was added relative lock time methods were explicitly out of scope.

File: src/psbt/mod.rs
Comment (tcharding) [reply src/psbt/mod.rs:354]: Sorry mate I'm not following you, which other methods?

File: src/miniscript/satisfy.rs
Comment (tcharding) [reply src/miniscript/satisfy.rs]: I've reverted these changes, so we now have
```rust
        // We need a relative lock time type in rust-bitcoin to clean this up.

        /* If nSequence encodes a relative lock-time, this mask is
         * applied to extract that lock-time from the sequence field. */
        const SEQUENCE_LOCKTIME_MASK: u32 = 0x0000ffff;
        const SEQUENCE_LOCKTIME_TYPE_FLAG: u32 = 0x00400000;

        let mask = SEQUENCE_LOCKTIME_MASK | SEQUENCE_LOCKTIME_TYPE_FLAG;
        let masked_n = n.to_consensus_u32() & mask;
        let masked_seq = self.to_consensus_u32() & mask;
        if masked_n < SEQUENCE_LOCKTIME_TYPE_FLAG && masked_seq >= SEQUENCE_LOCKTIME_TYPE_FLAG {
            false
        } else {
            masked_n <= masked_seq
        }
```

File: src/miniscript/satisfy.rs
Comment (LLFourn) [inline src/miniscript/satisfy.rs:109]: If you introduce a relative locktime type then it should probably be passed in here instead of `Sequence`.

File: src/miniscript/satisfy.rs
Comment (tcharding) [reply src/miniscript/satisfy.rs]: Please see https://github.com/rust-bitcoin/rust-miniscript/pull/455 which implements this using a new `relative::LockTime` API.

File: src/miniscript/satisfy.rs
Comment (tcharding) [reply src/miniscript/satisfy.rs:109]: Thanks man, I've done that now in the demo PR.

File: src/miniscript/astelem.rs
Comment (tcharding) [reply src/miniscript/astelem.rs:624]: I think I changed them all.

File: src/miniscript/types/mod.rs
Comment (tcharding) [reply src/miniscript/types/mod.rs:441]: I believe so.

Comment (tcharding) [reply src/miniscript/types/mod.rs:441]: I'm still at a loss as to exactly what the comment means. It should be removed now though, right?

**Comment by sanket1729** [comment]
I don't think we should wait for relative lock time PR in bitcoin 0.30.0. We would need a version of rust-miniscript that supports bitcoin 0.29.0

File: src/psbt/mod.rs
Comment (sanket1729) [reply src/psbt/mod.rs:354]: Oops. I meant I really like the new `added` methods upstream. They make the code here a lot cleaner. There is nothing to resolve here

File: src/policy/semantic.rs
Comment (sanket1729) [inline src/policy/semantic.rs:601]: This is okay, but I would have expected an API like 
```rust
let cmp_res = t.is_greater_than(n) ; // Any other name.  

match(cmp_res) {
	Some(true),
	Some(false),
	None,
}
```

Rust miniscript/or any other downstream project should need to know that they should check !t.is_same_unit() before comparing. 

Perhaps, something worth looking at 0.30.0

File: src/psbt/mod.rs
Comment (tcharding) [reply src/psbt/mod.rs:354]: :) Will be even better when `relative::LockTime` is available, its been nice to see the API in rust-bitcoin develop huh.

File: src/policy/semantic.rs
Comment (tcharding) [reply src/policy/semantic.rs:601]: We could add a method `is_satisfied_by_lock` in the same manner as `relative::LockTime`.

Comment (tcharding) [reply src/policy/semantic.rs:601]: Which would give us the much nicer:
```rust
            Policy::After(t) => {
                let t = LockTime::from(t);
                } else if !t.is_satisfied_by_lock(n) {
                    Policy::Unsatisfiable
                } else {
                    Policy::After(t.into())
                }
            }
```

Comment (tcharding) [reply src/policy/semantic.rs:601]: https://github.com/rust-bitcoin/rust-bitcoin/pull/1258

Comment (tcharding) [reply src/policy/semantic.rs]: Face palm, we actually provide a patter for this in the rustdocs (double face palm, I even wrote the rustdocs and still forgot to use the correct pattern)

I've updated the PR to be 
```rust
            Policy::After(t) => {
                let t = LockTime::from(t);
                let is_satisfied_by = match (t, n) {
                    (Blocks(t), Blocks(n)) => t <= n,
                    (Seconds(t), Seconds(n)) => t <= n,
                    _ => false
                };
                if !is_satisfied_by {
                    Policy::Unsatisfiable
                } else {
                    Policy::After(t.into())
                }
            }
```

**Comment by tcharding** [comment]
Hey @sanket1729, seems @josediegorobles is pretty keen to use the upcoming release, is there anything else that needs doing before we can consider this PR and get a release out?

**Comment by sanket1729** [comment]
Hi @tcharding, there is nothing that needs to be done in particular here. I have a few cleanups in mind, but those can wait for `8.1.0` . I will look into preparing a new release this week.

**Comment by josediegorobles** [comment]
Hi, @sanket1729 by suggestion of @tcharding i started a PR for the release: https://github.com/tcharding/rust-miniscript/pull/1

What version must be, 7.1.0 or 8.0.0? And what other things must be on the changelog?

**Review by sanket1729 — APPROVED** [review]
ACK 24a249820f403ec103ab7e3d45fa4f7a5f0dee19. Reviewed by range-diff from c51595b9c951d1f07b41dccb79d2205cc3c91160 .


I am waiting on @apoelstra to respond the comment issue he raised

File: src/miniscript/types/mod.rs
Comment (sanket1729) [reply src/miniscript/types/mod.rs:824]: @apoelstra are you convinced that this is resolved?

Comment (apoelstra) [reply src/miniscript/types/mod.rs:824]: Let me re-read this PR.

File: src/policy/semantic.rs
Comment (apoelstra) [inline src/policy/semantic.rs:370]: nit: there was no need for the `|x|` here

**Review by apoelstra — APPROVED** [review]
ACK 24a249820f403ec103ab7e3d45fa4f7a5f0dee19

**Comment by apoelstra** [comment]
@sanket1729 in future before merging can you edit PR descriptions to remove the `@` signs before people's names? Otherwise github creates spurious notifications for them.

**Comment by tcharding** [comment]
Should I just not put mentions in the PR description? I thought it was only in the commit log (on the actual patch) that mattered?

**Comment by apoelstra** [comment]
@tcharding no, it matters both on the commits and in the PR description. (Though the PR description is really easy to edit, for anyone with merge permissions, while the commits are a pain.)

**Comment by tcharding** [comment]
My bad, duly schooled :)

**Comment by Kixunil** [comment]
@apoelstra are you sure? I was never spuriously notified from PRs. Creation doesn't count as spurious - I consider mentioning explicitly wanting to notify someone.

**Comment by apoelstra** [comment]
@Kixunil yes, though I believe the reason is that our merge script copies the entire PR description into the merge commit message.

**Comment by Kixunil** [comment]
Ah, good point! Could the merge script be modified to replace `([:whitespace:])@([a-zA-Z0-9_-]+[:whitespace]])` with `\1\2` in PR description?

**Comment by apoelstra** [comment]
I don't know -- could you open an issue on https://github.com/bitcoin-core/bitcoin-maintainer-tools/ ?

Note that it does currently display an angry warning if there are @s in the PR description, which I always respond to by just editing the PR description and trying again ... maybe they don't want to modify PRs by default but it'd be nice if there was at least a gitconfig flag or something to do it.

**Pull request merged**
